R/keyword_quote_linter.R

```diff
@@ -0,0 +1,129 @@
+#' Block unnecessary quoting in calls
+#'
+#' Any valid symbol can be used as a keyword argument to an R function call.
+#'   Sometimes, it is necessary to quote (or backtick) an argument that is
+#'   not an otherwise valid symbol (e.g. creating a vector whose names have
+#'   spaces); besides this edge case, quoting should not be done.
+#'
+#' The most common source of violation for this is creating named vectors,
+#'   lists, or data.frame-alikes, but it can be observed in other calls as well.
+#'
+#' @evalRd rd_tags("keyword_quote_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+# TODO(michaelchirico): offer a stricter version of this that
+#   requires backticks to be used for non-syntactic names (i.e., not quotes).
+#   Here are the relevant xpaths:
+#   //expr[expr[SYMBOL_FUNCTION_CALL]]/SYMBOL_SUB[starts-with(text(), '`')]
+#   //expr[expr[SYMBOL_FUNCTION_CALL]]/STR_CONST[{is_quoted(text())}]
+keyword_quote_linter <- function() {
+  # NB: xml2 uses xpath 1.0 which doesn't support matches() for regex, so we
+  #   have to jump out of xpath to complete this lint.
+  # It's also a bit tough to get the escaping through R and then xpath to
+  #   work as intended, hence the rather verbose declaration here.
+  quote_cond <- xp_or(
+    "starts-with(text(), '\"')",
+    "starts-with(text(), '`')",
+    'starts-with(text(), "\'")'
+  )
+  # SYMBOL_SUB for backticks, STR_CONST for quoted names
+  call_arg_xpath <- glue("
+  //SYMBOL_FUNCTION_CALL
+    /parent::expr
+    /parent::expr
+    /*[(self::SYMBOL_SUB or self::STR_CONST) and {quote_cond}]
+  ")
+
+  # also exclude $ or @, which are handled below
+  assignment_xpath <- "
+  (//EQ_ASSIGN | //LEFT_ASSIGN[text() != ':='])
+    /preceding-sibling::expr[
+      not(OP-DOLLAR or OP-AT)
+      and (STR_CONST or SYMBOL[starts-with(text(), '`')])
+    ]
+  "
+
+  extraction_xpath <- "
+    (//OP-DOLLAR | //OP-AT)/following-sibling::STR_CONST
+    | //OP-DOLLAR/following-sibling::SYMBOL[starts-with(text(), '`')]
+    | //OP-AT/following-sibling::SLOT[starts-with(text(), '`')]
+  "
+
+  return(Linter(function(source_expression) {
+    if (!is_lint_level(source_expression, "expression")) {
+      return(list())
+    }
+
+    xml <- source_expression$xml_parsed_content
+
+    call_arg_expr <- xml_find_all(xml, call_arg_xpath)
+
+    invalid_call_quoting <- is_valid_r_name(get_r_string(call_arg_expr))
+
+    call_arg_lints <- xml_nodes_to_lints(
+      call_arg_expr[invalid_call_quoting],
+      source_expression = source_expression,
+      lint_message = "Only quote named arguments to functions if necessary, i.e., the name is not a valid R symbol (see ?make.names).",
+      type = "warning"
+    )
+
+    assignment_expr <- xml_find_all(xml, assignment_xpath)
+
+    invalid_assignment_quoting <- is_valid_r_name(get_r_string(assignment_expr), no_quote = TRUE)
+
+    assignment_lints <- xml_nodes_to_lints(
```

Comment:
I think it's worth splitting the cases here

> Use ` for quoting names, not %s.

And the other message for unnecessary quoting can be same as above.